### PR TITLE
add apollo fix graphiql

### PR DESCRIPTION
**Pull request by davlee1972**
Added Apollo Client
Added window.location.href as initialEndpoint
Added default_query option to Apollo Client

Fixed GraphiQL Client - Was stuck on loading screen. The default query string injection was failing.
Added window.location.href as url for fetcher()
Added default_query option to GraphiQL Client

I couldn't figure out how to get double quotes to render properly.. `&quot; `is showing up instead of "

```
# GraphQL queries typically start with a &quot;{&quot; character.Lines that start
# with a # are ignored.
#
# An example GraphQL query might look like:
#
#     {
#       field(arg: &quot;value&quot;) {
#         subField
#
```

### CODE REVIEW

**Review by rafalp — COMMENTED** [review]
We also have tests that explorers work for ASGI app and WSGI app.

**Comment by davlee1972** [comment]
Ok. I'll take a look at the other unit tests. Still figuring out how this snapshot library works..

The other problem is with rendering templates with tokens. It doesn't like substituting a string which contains a string.

I might have to just do a string replace after:

GRAPHIQL_HTML = read_template("graphiql.html")

```
# Create an ASGI app using the schema, running in debug mode
app = GraphQL(schema, debug=True, explorer = ExplorerGraphiQL(default_query="""{people ("firstName": "Bob"){fullName}}"""))
```

Is rendering as:

`{people (&quot;firstName&quot;: &quot;Bob&quot;){fullName}}
`

**Comment by rafalp** [comment]
There's `{% raw var_name %}` that prints variable in template without escaping, but you will still need to escape string for JavaScript. I've didn't implement special logic for this and IMHO it should be an utility function instead of yet another template tag.

File: requirements-dev.in [hunk 1]
```diff
@@ -5,15 +5,15 @@ mypy
...
+pytest - mock
+python - dateutil
+python - multipart
 requests
 snapshottest
-types-python-dateutil
+types - python - dateutil
 werkzeug
 
-pip-tools
```
Comment (rafalp) [inline requirements-dev.in:19]: Change out of scope?

**Comment by davlee1972** [comment]
Added ASGI and WSGI explorer tests
Updated anapshots
Used {% raw variable_name %} to add default_query

File: requirements-dev.in (hunk 1 above)
Comment (davlee1972) [reply requirements-dev.in:19]: I restored requirements-dev.in... Not sure how it got changed.

File: requirements-dev.in [hunk 2]
```diff
@@ -16,4 +16,4 @@ snapshottest
 types-python-dateutil
 werkzeug
 
-pip-tools
+pip-tools
```
Comment (rafalp) [inline requirements-dev.in:19]: Also change out of scope

File: tests/asgi/snapshots/snap_test_explorer.py [hunk 3]
```diff
@@ -18,96 +35,69 @@
...
-        fetcher: fetcher,
-        defaultEditorToolsVisibility: true,
-        query: query,
-      }),
-      document.getElementById('graphiql'),
-    );
+  </head>
+
+  <body>
+    <div id="graphiql">Loading...</div>
```
Comment (rafalp) [inline tests/asgi/snapshots/snap_test_explorer.py:58]: Please take a note that we've included `{{ title }}` in the loading message.

File: tests/asgi/test_explorer.py [hunk 4]
```diff
@@ -13,6 +14,15 @@ def test_default_explorer_html_is_served_on_get_request(schema, snapshot):
     client = TestClient(app)
     response = client.get("/")
     assert response.status_code == 200
+    print(response.text)
```
Comment (rafalp) [inline tests/asgi/test_explorer.py:17]: Print 🚨

**Review by rafalp — CHANGES_REQUESTED** [review]
I've left few notes.

File: tests/asgi/snapshots/snap_test_explorer.py (hunk 3 above)
Comment (davlee1972) [reply tests/asgi/snapshots/snap_test_explorer.py]: added {{ title }}

File: requirements-dev.in (hunk 2 above)
Comment (davlee1972) [reply requirements-dev.in]: Not sure why this is popping up. There isn't a real change to this file

File: tests/asgi/test_explorer.py (hunk 4 above)
Comment (davlee1972) [reply tests/asgi/test_explorer.py]: removed pring

**Review by rafalp — APPROVED** [review]

**Pull request merged**